Convert multiinputfile to use context / core

kdt3rd opened the pull request.

kdt3rd committed:
Fix grammar

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Fix speling

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
remove tracking of current position

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Improve exception messages

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Add test, accessor to match multipart behavior

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Clean extraneous deep data during chunk reconstruction

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
fix checkfile typo, return value for ripmap vs mipmap tiles

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Convert MultiPartInputFile to use context

Also simplifies InputFile to always defer to MultiPartInputFile to start
Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
remove extra spaces

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

meshula commented on src/lib/OpenEXR/ImfContextInit.cpp:
```diff
@@ -17,13 +17,11 @@ struct istream_holder
 #if ILMTHREAD_THREADING_ENABLED
     std::mutex _mx;
 #endif
-    uint64_t _cur_offset = 0;
```
I see the incremental removal of statefulness :)

meshula commented on src/lib/OpenEXR/ImfDeepTiledInputFile.cpp:
```diff
@@ -1861,7 +1861,7 @@ DeepTiledInputFile::readPixelSampleCounts (
                     if (partNumber != _data->partNumber)
                         throw IEX_NAMESPACE::InputExc (
-                            "Unexpected part number.");
+                            "Unexpected part number for tile offset.");
```
```suggestion
                            "Unexpected part number for tile offset." + std::to_string(partNumber));
```

meshula commented on src/lib/OpenEXR/ImfInputFile.h:
```diff
@@ -41,13 +38,13 @@ class TiledInputFile;
 /// MultiPartInputFile
 ///   can be constructed but not directly accessible
-///   - InputPart
+///   - InputPart (the part analog to this class)
```
```suggestion
///   - InputPart (the part corresponding to this object)
```
?

meshula commented on src/lib/OpenEXR/ImfInputFile.h:
```diff
@@ -118,8 +110,8 @@ class IMF_EXPORT_TYPE InputFile : public GenericInputFile
 
-    OPENEXR_DEPRECATED (
-        "Use context-based attribute access for faster retrieval")
+    //OPENEXR_DEPRECATED (
```
the deprecation attribute is SUPER annoying because -Wall causes SO much spam. What if we added to the cmake, "EXR_WARN_ON_DEPRECATED=OFF" and predicate the OPENEXR_DEPRECATED macro on that? deprecation warnings are useful, but I'd only want to turn them on briefly when migrating to a new version of the library

kdt3rd replied:
yeah, not 100% sure what to do with that, I feel like I want to put them in there prior to release, but you're right, we should think about a convenient enough way for a dev. to silence the deprecations if they can't change things just yet...

lgritz replied:
That's easy to do in the downstream project, it's a specific warning

lgritz replied:
But we could also have a magic symbol that if set before inclusion of the exr headers, just defines all the DEPRECATED macros to do nothing.

meshula replied:
The thing about setting it downstream is that that becomes all or nothing; you might want your own deprecation warnings, but not the ones out of your control. For this reason, it's a fairly usual pattern to allow suppression of a project's deprecations via a macro.

lgritz replied:
That's what I meant. If you set OPENEXR_NO_DEPRECATION_WARNINGS prior to including the openexr headers, you won't get the warnings. (I mean, we'd need to honor that.)

meshula commented on src/lib/OpenEXR/ImfScanLineInputFile.cpp:
```diff
@@ -373,7 +373,7 @@ readPixelData (
     //
 
-    if (!isMultiPart (ifd->version))
+    if (ifd->partNumber == -1 && !isMultiPart (ifd->version))
```
is -1 a magic sentinel, or does this want to be < 0?

kdt3rd replied:
it is magic to the c++ land (in the constructor for the data object), this stuff will go away soon enough :)

meshula commented on src/lib/OpenEXR/ImfTiledInputFile.cpp:
```diff
@@ -1439,7 +1438,7 @@ TiledInputFile::rawTileData (
         int old_lx = lx;
         int old_ly = ly;
-        if (isMultiPart (version ()))
+        if (_data->partNumber != -1 || isMultiPart (version ()))
```
ditto on sentinel, I'm not quite sure where the value originates

meshula reviewed (commented):
this looks good to me; although I'm wondering if the deprecation macro can be controlled via cmake rather than ccommented out

kdt3rd committed:
Reproduce check against the count of chunk items

The C++ library had arbitrary (duplicate, non-editable) checks against a
max chunk count, because the IO streams can't always provide the file
size as a safety check. reproduce that here

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Address review notes from meshula

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Small cleanup

reduce scope of variables, collapse logic to avoid duplicate conditional
blocks, re-enable failable free test (original issues had already been
fixed, but test had not been re-enabled)

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Remove commented deprecations until further review

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
add try / catch to avoid throwing through the C interface

Streams that throw are an unfortunate choice, but such is the way.

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>
